### PR TITLE
Strongly type bootstrap interface

### PR DESCRIPTION
**Pull request by nmittler**
The bootstrap package is loosely typed and it's not clear what values are settable or what their types are.

This PR changes the bootstrap package to clearly indicate what parameters it supports as well as their types.

This is a step toward a broader refactoring of the pilot agent code.

Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
quick skin looks good thanks for cleaning this up

I'm curious why you went with this approach (is this what would be called "functional arguments"?) rather than a huge struct. Both seem like they would work, so just curious about pros/cons ?

**Comment by rlenglet** [comment]
> This is a step toward a broader refactoring of the pilot agent code.

Can you link to any issue / doc detailing what other refactoring you're planning to do?

**Comment by hzxuzhonghu** [comment]
It makes the code harder to comprehend than before, convert map to json or yaml is a usual pattern. We donot need so abstract wrapper.   If we really want to do some cleanup, i think separate it into different module is ok.

**Comment by nmittler** [comment]
@howardjohn 

> I'm curious why you went with this approach (is this what would be called "functional arguments"?) rather than a huge struct. Both seem like they would work, so just curious about pros/cons ?

Having a monolithic config object is a nice way to gather everything into one place ... I'm not entirely against it.  However in this case, the config object would need a long conversion method that converts each option. Breaking the options out into separate objects allows us to more cleanly combine the interface with the conversion logic for each option.

**Comment by nmittler** [comment]
@hzxuzhonghu

> It makes the code harder to comprehend than before, convert map to json or yaml is a usual pattern. We donot need so abstract wrapper. If we really want to do some cleanup, i think separate it into different module is ok.

I strongly disagree. Using a basic map with arbitrary values is incredibly error prone and hard to reason about as a user without knowing all of the details of the template. This change makes the API static and much more strongly typed, so it's more clear which options are available and their types.

**Comment by nmittler** [comment]
@rlenglet 

> Can you link to any issue / doc detailing what other refactoring you're planning to do?

I don't have a doc as of yet. I've been doing a lot of this as a background task around other work. Most of this so far has been aimed cleaning up `/pkg/envoy`. I have #17355 out for review, but there are some open comments that might suggest further simplification. The change in this PR to `pkg/bootstrap` is also related to that work and will help to further simplify it. In addition, the bootstrap code in `pkg/bootstrap` really belongs in `pkg/envoy/bootstrap` (I'm planning on doing that move after this PR lands).

As far as the rest of pilot-agent ... I think it's fair to say there is a lot of work there. Definitely worthy of a doc, and will likely require involvement from a few motivated folks.

File: pkg/bootstrap/instance.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+func newTemplate(config *meshAPI.ProxyConfig) (*template.Template, error) {
+	templateFilePath := config.CustomConfigFile
+	if templateFilePath == "" {
+		templateFilePath = config.ProxyBootstrapTemplatePath
+	}
+	if templateFilePath == "" {
+		templateFilePath = DefaultCfgDir
+	}
```
Comment (sbezverk) [inline pkg/bootstrap/instance.go]: It looks to me a bit more readable this way:

switch {
case config.CustomConfigFile != "": 
         templateFilePath := config.CustomConfigFile
case config.ProxyBootstrapTemplatePath != "":
         templateFilePath = config.ProxyBootstrapTemplatePath
default: 
         templateFilePath = DefaultCfgDir
}

**Comment by howardjohn** [comment]
> However in this case, the config object would need a long conversion method that converts each option

I am not sure what you mean by this. Do you mean we would need to do something like 
`opts.SomeField = convertSomeField(something)`? rather than your approach where this in its own method in `options.go` and we just do `opts.set(SomeField(something))`?

**Comment by kyessenov** [comment]
The problem with applying static type here is that we don't control the declarations. These are a part of xDS APIs and change frequently. Putting static types in place makes adapting to changes in xDS harder, especially since we don't follow cross-versioning systematically (e.g. which Envoys would istio 1.3.1 work with?)

**Comment by nmittler** [comment]
@kyessenov 

> The problem with applying static type here is that we don't control the declarations. These are a part of xDS APIs and change frequently. Putting static types in place makes adapting to changes in xDS harder, especially since we don't follow cross-versioning systematically (e.g. which Envoys would istio 1.3.1 work with?)

Could you expand on how adapting to Envoy version changes would be easier if we just supported a raw map?

**Comment by kyessenov** [comment]
@nmittler it would be easier to apply break-glass fixes with dynamic scripting (e.g. go templates).

As an example, we had to modify mixer bootstrap recently to apply request-level load balancing for mixer adapters. This would be harder to achieve if the bootstrap was rigidly schematized since we would likely miss some aspect of xDS that just was not there when we release istio.

It's the same reason why helm does not try to schematize k8s APIs. The API surface is huge, and it changes after you release the binary.

**Comment by nmittler** [comment]
@kyessenov I'm not sure I fully understand the problem. If the break-glass change would be to modify the go template file, that will be unaffected by this change ... the map that's passed in is identical to the one before. Or am I missing something?

**Comment by kyessenov** [comment]
@nmittler So if the consumer takes a map to compute the template, and the user writes a map using environment variables (and just modifying the template), what's the value of schematizing the bits in the middle?

**Comment by nmittler** [comment]
@kyessenov 

> So if the consumer takes a map to compute the template, and the user writes a map using environment variables (and just modifying the template), what's the value of schematizing the bits in the middle?

I'm a little uncertain of what you mean by `consumer` and `user` in this context.  The purpose of the strong typing is to make a clear API surface for the library, identifying what options are settable and providing reasonable types where it makes sense. Internally, calls on this API are transformed into template parameters ... the same ones as before.

File: pilot/cmd/pilot-agent/main.go [hunk 2]
```diff
@@ -320,43 +320,34 @@ var (
...
+					opts = append(opts, bootstrap.PodNamespace(podNamespaceVar.Get()))
+
+					// protobuf encoding of IP_ADDRESS type
+					opts = append(opts, bootstrap.PodIP(net.ParseIP(instanceIPVar.Get())))
+
 					mixerSAN := getSAN(ns, envoyDiscovery.MixerSvcAccName, mixerIdentity)
 					log.Infof("MixerSAN %#v", mixerSAN)
 					if len(mixerSAN) > 1 {
-						opts["MixerSubjectAltName"] = mixerSAN[0]
+						opts = append(opts, bootstrap.MixerSubjectAltName(mixerSAN[0]))
```
Comment (rshriram) [inline pilot/cmd/pilot-agent/main.go]: you can get rid of the mixerSAN because sidecars do not need this

**Review by howardjohn — APPROVED** [review]

**Comment by howardjohn** [comment]
I suspect something went wrong here:
```
    mixer_test.go:712: Unexpected 'unknown' value for label 'destination_workload_uid' in sample 'istio_kube_request_count{destination_container="unknown", destination_owner="unknown", destination_pod="unknown", destination_workload="unknown", destination_workload_uid="unknown", instance="10.244.0.12:42422", job="istio-mesh", response_code="200", source_owner="kubernetes://apis/apps/v1/namespaces/istio-system/deployments/details-v1", source_pod="details-v1-75f77b97b7-clwcz", source_workload="details-v1", source_workload_uid="istio://istio-system/workloads/details-v1"} => 2 @[1569964395.369]'

```

You can see it has failed like 10 times https://prow.istio.io/pr-history/?org=istio&repo=istio&pr=17478

**Comment by kyessenov** [comment]
@nmittler What I mean is that this makes it difficult to extend bootstrap template with new options. For example, opencensus config was added to bootstrap but did no yet propagate here. So now we need to add a lot of boilerplate code just to specify opencensus via environment variables.

**Comment by nmittler** [comment]
> @nmittler What I mean is that this makes it difficult to extend bootstrap template with new options. For example, opencensus config was added to bootstrap but did no yet propagate here. So now we need to add a lot of boilerplate code just to specify opencensus via environment variables.

One way or another the properties have to get fished through to the template via real code. This just enforces clear semantics on the application-facing interface.

**Pull request merged**